Semantics.md

```diff
@@ -683,3 +683,19 @@ outside the range which rounds to an integer in range) traps.
 [future types]: FutureFeatures.md#more-table-operators-and-types
 [future large pages]: FutureFeatures.md#large-page-support
 [future ieee 754]: FutureFeatures.md#full-ieee-754-2008-conformance
+
+
+## Validation
+
+A module binary has to be _validated_ before it is compiled.
+Validation ensures that the module is well-defined and that its code cannot exhibit any undefined behaviour.
+In particular, along with some runtime checks, it ensures that no program can access or corrupt memory it does not own.
+
+Validation of code is mostly defined in terms of type-checking the use of the operand stack.
+It sequentially checks for each that the expected operands can be popped from the stack, and tracks which new operands are pushed onto it. At the start of a function the stack is empty; at its end it has to match the return type of the function. In addition, instructions inside a `block` (or `loop` or `if`) cannot consume operands pushed outside; at the end of the block the remaining inner operands must match the block signature.
+
+A special case are instructions performing an unconditional control flow transfer (`br`, `br_table`, `return`, `unreachable`), because execution never proceeds after them.
+The stack after such an instruction is defined as _unreachable_.
+An unreachable stack is valid input for any instruction, and an instruction with unreachable input will produce unreachable output.
+An unreachable stack also matches any possible signature at the end of a block or function.
+After the end of a block, the stack is determined by the block signature and the stack before the block.
```

Comment:
Maybe this could be clarified a bit, as it currently reads backward to me. How about "After the end of a block, the stack contains the contents of the stack before the block with the block's signature appended."